src/v3/src/util/deviceFingerprintingUtils.ts

```diff
@@ -0,0 +1,102 @@
+
+export const isMessageFromCorrectSource = (iframe: HTMLIFrameElement, event: MessageEvent)
+: boolean => (
+  event.source === iframe.contentWindow
+);
+
+// NOTE: This utility is similar to the DeviceFingerprinting.js file used for V2 authentication flows.
+export const generateDeviceFingerprint = (oktaDomainUrl: string): Promise<string> => (
+  // eslint cannot detect that setTimeout() at end of function will eventually return with Error rejection
+  // eslint-disable-next-line consistent-return
+  new Promise((resolve, reject) => {
+    const userAgent = getUserAgent();
+    if (!userAgent) {
+      return reject(new Error('User agent is not defined'));
+    } if (isWindowsPhone(userAgent)) {
+      return reject(new Error('Device fingerprint is not supported on Windows phones'));
+    }
+
+    let iframe: HTMLIFrameElement;
+    let iframeTimeout: NodeJS.Timeout;
+    let onMessageReceivedFromOkta: (event: MessageEvent) => void;
+
+    const removeIframe = () => {
+      iframe.remove();
+      window.removeEventListener('message', onMessageReceivedFromOkta);
+    };
+
+    const handleError = (reason: string) => {
+      removeIframe();
+      return reject(new Error(reason));
+    };
+
+    const sendMessageToOkta = (message: { type: string }) => {
+      const win = iframe.contentWindow;
+
+      if (win) {
+        win.postMessage(JSON.stringify(message), oktaDomainUrl);
+      }
+    };
+
+    onMessageReceivedFromOkta = (event: MessageEvent) => {
+      if (!isMessageFromCorrectSource(iframe, event)) {
+        return;
+      }
+      // deviceFingerprint service is available, clear timeout
+      clearTimeout(iframeTimeout);
+      if (!event || !event.data || event.origin !== oktaDomainUrl) {
+        handleError('No data');
+      }
+      try {
+        const message = JSON.parse(event.data);
+
+        if (message && message.type === 'FingerprintServiceReady') {
+          sendMessageToOkta({
+            type: 'GetFingerprint',
+          });
+        } else if (message && message.type === 'FingerprintAvailable') {
+          removeIframe();
+          resolve(message.fingerprint);
+        } else {
+          handleError('No data');
+        }
+      } catch (error) {
+        // Ignore any errors since we could get other messages too
+      }
+    };
+
+    // Attach listener
+    window.addEventListener('message', onMessageReceivedFromOkta, false);
+    iframe = document.createElement('iframe');
+    iframe.id = 'device-fingerprint-container';
+    iframe.style.display = 'none';
+    // Create and load devicefingerprint page inside the iframe
+    iframe.src = `${oktaDomainUrl}/auth/services/devicefingerprint`;
+
+    const formElement = document.querySelector('form[data-se="o-form"]');
+    if (formElement === null) {
+      handleError('Form does not exist');
+    }
+    formElement!.appendChild(iframe);
```

Comment:
Gen2 utility uses jQuery to create the `iframe`, switched to dynamically creating `iframe` and adding it to the DOM. Also, gen2 utility was passed the `form` element so `iframe` can be appended, but we do not have access to that `form` element in `useOnSubmit` so we search for the `form` using `data-se="o-form"` instead